src/cmd/input.go

```diff
@@ -30,34 +34,108 @@ func readInputConfig() {
 	viper.ReadInConfig()
 }
 
-func readResourceInput[T any]() (*T, error) {
+func readInput() ([]byte, error) {
 	var err error
-	var resource T
-	var yamlData []byte
+	var input []byte
 
 	switch dataFile {
 	case ".":
-		yamlData, err = os.ReadFile("./data.yaml")
+		input, err = os.ReadFile("./data.yaml")
 	case "-":
 		if isStdInFromTerminal() {
 			log.Info().Msg("Reading input directly from command line... Press CTRL+D to stop typing")
 		}
 		buf := bytes.Buffer{}
 		_, err = buf.ReadFrom(os.Stdin)
-		yamlData = buf.Bytes()
+		input = buf.Bytes()
 	default:
-		yamlData, err = os.ReadFile(dataFile)
+		input, err = os.ReadFile(dataFile)
 	}
 	if err != nil {
-		return nil, err
+		return input, err
 	}
+	return input, nil
+}
 
-	if err := yaml.Unmarshal(yamlData, &resource); err != nil {
+func ReadResource[T any](input []byte) (*T, error) {
+	var resource T
+	if err := yaml.Unmarshal(input, &resource); err != nil {
 		return nil, err
 	}
 	return &resource, nil
 }
 
+func handleJSONSchema(m map[string]any) error {
+	// since a YAML object is map[string]any which is compatible with JSONSchema so nothing needs to be done.
+	// if schema is a string, the JSON inside the string needs to be parsed. supports multiline.
+	if _, ok := m["schema"].(string); !ok {
+		return nil
+	}
+	jsonSchema, err := opslevel.NewJSONSchema(m["schema"].(string))
+	if err != nil {
+		return err
+	}
+	m["schema"] = jsonSchema
+	return nil
+}
+
+func handleJSONString(m map[string]any) error {
+	// TODO: multiline JSON is not supported as a value input because handling whitespace and escape characters is nontrivial
+	jsonString, err := opslevel.NewJSONInput(m["value"])
+	if err != nil {
+		return err
+	}
+	m["value"] = jsonString
+	return nil
+}
+
+func ReadResourceHandleJSONFields[T any](input []byte) (*T, error) {
```

Comment:
It looks like this is ONLY used in test files - this shouldn't be part of the codebase then it should be a utility function in the testing code.

Reply:
Same with its helper functions `handleJSONString` and `handleJSONSchema`

Reply:
@rocktavious this is used in the properties PR that follows this: https://github.com/OpsLevel/cli/pull/257

Reply:
In fact after reading the whole code - i don't see where this function is used - so why are we testing it?

Reply:
@rocktavious I wanted to add an abstraction in case we would start adding more resource types that use JSON fields outside of properties, but I have now reversed this.

Reply:
I feel like the abstraction is more complex and unexpected then having custom yaml marshallars

Reply:
@rocktavious if we remove support for JSON, we could replace this code with only custom YAML marshalers which would be much cleaner. Supporting both YAML and JSON feels hacky. 

Also, while you can read in JSON objects, reading in JSON values (number, string, boolean, list) is next to impossible without us writing another 100+ lines of code. 

How do you feel about only supporting YAML?

Reply:
https://github.com/OpsLevel/cli/pull/255#discussion_r1514876000